torch/csrc/jit/tensorexpr/ir_simplifier.h

```diff
@@ -0,0 +1,542 @@
+      val = imm->value();                                              \
+    } else {                                                           \
+      LOG(FATAL) << "Bad expr: " << *expr << "\n";                     \
+    }                                                                  \
+    break;
+    AT_FORALL_SCALAR_TYPES_AND2(Half, Bool, TYPE_CASE);
+#undef TYPE_CASE
+    default:
+      LOG(FATAL) << "Unsupported datatype: " << expr->dtype();
+  }
+
+  return val;
+}
+
+/* Takes a LinearForm and converts it to Mul + (Add/Sub). */
+const Expr* expandLinearForm(const LinearForm* v, IRMutator* mutator) {
+  const Expr* mul = nullptr;
+  const Expr* A = v->getA();
+  const Expr* B = v->getB();
+  // X can be a LinearForm.
+  const Expr* X = v->getX()->accept_mutator(mutator);
+  // we only really care about 0 and 1, so double should be fine.
+  double Aval = immediateAs<double>(A);
+  double Bval = immediateAs<double>(B);
+
+  // First handle A.
+  if (Aval == 0) {
+    if (Bval == 0) {
+      return getImmediateByType(X->dtype(), 0).node();
+    }
+    return B;
+  } else if (Aval == 1) {
+    mul = X;
+  } else if (Aval == -1) {
+    return new Sub(B, X);
+  } else if (Aval < 0) {
+    // Negate A.
+    ExprHandle zero = getImmediateByType(A->dtype(), 0);
+    Sub* A_Sub = new Sub(zero.node(), A);
+
+    return new Sub(B, new Mul(X, evaluateOp(A_Sub)));
+  } else {
+    mul = new Mul(X, A);
+  }
+
+  if (Bval == 0) {
+    return mul;
+  }
+
+  return new Add(mul, B);
+}
+
+/* Expand any remaining LinearTerms into their component pieces */
+class TermExpander : public IRMutator {
+ public:
+  const Expr* mutate(const LinearForm* v) {
+    return expandLinearForm(v, this);
+  }
+};
+
+/* Simplify the IR by combining arithmetic expressions over a common term.
+ */
+class IRSimplifier : public IRMutator {
+ public:
+  /* Expands lhs and rhs if they are LinearTerms, recurses if they can be folded
+   * into a constant form. */
+  const Expr* expandAndRecurse(
+      const Expr* expr,
+      const Expr* lhs,
+      const Expr* rhs) {
+    if (const LinearForm* lhsLinear = dynamic_cast<const LinearForm*>(lhs)) {
+      lhs = expandLinearForm(lhsLinear, this);
+    }
+    if (const LinearForm* rhsLinear = dynamic_cast<const LinearForm*>(rhs)) {
+      rhs = expandLinearForm(rhsLinear, this);
+    }
+    const Expr* result = newBinaryOpOfType(expr->expr_type(), lhs, rhs, false);
+
+    if (lhs->isConstant() || rhs->isConstant()) {
+      // One wasn't really a linear, we can redo this.
+      return result->accept_mutator(this);
+    }
+
+    return result;
+  }
+
+  /* Handles optimization cases for Broadcast() + Other */
+  const Expr* handleBroadcastAdd(const Broadcast* bc, const Expr* other) {
+    if (bc->value()->isConstant() && immediateAs<int>(bc->value()) == 0) {
+      return other;
+    }
+
+    if (const Ramp* r = dynamic_cast<const Ramp*>(other)) {
+      const Expr* ret = Ramp::make(
```

Comment:
Nit: you can do `new Ramp(new Add(bc->value(), r->base()), r->stride(), r->lanes())` to avoid boxing/unboxing.